### PR TITLE
[Performance][MRV2] Cache logits-processing request state

### PR DESCRIPTION
**Pull request by positive666**
## Summary

`Sampler._requires_logits_processing()` runs on every sampling step and currently scans multiple per-request NumPy/Python states. This change computes the combined predicate in `add_request()` and stores one boolean per request slot, so the hot-path gate only scans the active rows of one compact array.

The cache is overwritten on slot reuse, keeps active-subset filtering through the existing `idx_mapping_np`, and does not change any public or downstream processor API.

## Dependency

Depends on #52284, because an active thinking budget must participate in the combined predicate. This branch carries the rebased #52284 change as its first commit; after #52284 lands, rebasing will leave only the performance commit in this PR's diff.

## Why this is not duplicate work

There is no linked issue. Immediately before opening this PR I searched open PRs for `_needs_logits_processing`, `logits-processing sampler`, `MRV2 sampler cache`, and `52284 in:body`. No PR implements this cached combined predicate. #50584 adds DRY sampling state, and #48601 addresses FlashInfer top-k/top-p warmup OOM; neither is the same optimization.

### Unit and static checks

```bash
/tmp/vllm-pr2-venv/bin/python -m pytest \
  tests/v1/worker/test_gpu_sampler_flags.py \
  tests/v1/worker/test_gpu_thinking_budget.py -q
# 26 passed, 14 warnings in 18.42s

uv run --active pre-commit run ruff-check --files \
  vllm/v1/worker/gpu/sample/sampler.py \
  tests/v1/worker/test_gpu_sampler_flags.py \
  tests/v1/worker/test_gpu_thinking_budget.py
uv run --active pre-commit run ruff-format --files \
  vllm/v1/worker/gpu/sample/sampler.py \
  tests/v1/worker/test_gpu_sampler_flags.py \
  tests/v1/worker/test_gpu_thinking_budget.py
uv run --active pre-commit run mypy-3.12 --hook-stage manual --files \
  vllm/v1/worker/gpu/sample/sampler.py \
  tests/v1/worker/test_gpu_sampler_flags.py \
  tests/v1/worker/test_gpu_thinking_budget.py
# all passed
```

A broader relevant pre-commit run also passed all Python checks. `actionlint` could not download its dependency from `proxy.golang.org`, and `update-dockerfile-graph` requires `/bin/bash` on the Windows host; neither hook covers these Python files.

The focused tests cover the default and greedy no-op paths, thinking budget, logit bias, penalties, bad words, temperature, min-p, top-k, top-p, warmup, slot reuse, and active-subset filtering.

### Model behavior and quality

- Qwen3.5-9B fixed-seed A/B matrix: exact output equality for default, greedy, temperature, top-p, top-k, min-p, penalties, logit bias, bad words, and thinking budget; all requests returned HTTP 200.
- Repository GSM8K evaluator (`tests/evals/gsm8k/gsm8k_eval.py`), Qwen3.5-9B, 100-question sample, 5-shot greedy, seed 42, max 256 tokens:
  - baseline: accuracy 0.87, invalid rate 0, 10,644 output tokens
  - optimized: accuracy 0.87, invalid rate 0, 10,644 output tokens

### Performance

Host-side decision-path microbenchmark, 50,000 calls x 9 repeats:

| Batch | Baseline median | Optimized median | Change |
|---:|---:|---:|---:|
| 1 | 56.708 us | 16.830 us | -70.3% |
| 32 | 57.137 us | 16.971 us | -70.3% |
| 256 | 62.550 us | 18.908 us | -69.8% |

Qwen3-0.6B serving benchmark (4 baseline and 3 optimized runs, 1,024 requests/run, 32 input and 256 output tokens, concurrency 256):

```bash
vllm bench serve --backend openai --base-url http://127.0.0.1:8001 \
  --model Qwen3-0.6B --tokenizer /model --dataset-name random \
  --input-len 32 --output-len 256 --num-prompts 1024 --num-warmups 128 \
  --max-concurrency 256 --request-rate inf --ignore-eos --temperature 1.0 \
  --top-p 1.0 --top-k 0 --min-p 0.0 --disable-shuffle --seed 0 \
  --disable-tqdm
```

- median output throughput: 16,731.06 -> 17,488.81 tok/s (+4.53%)
- median mean TPOT: 13.713 -> 12.891 ms (-5.99%)
- every run completed 1,024/1,024 requests successfully
- a fourth baseline run after the optimized runs remained at 16,574.05 tok/s, checking for warmup/order bias

Qwen3.5-9B serving benchmark (3 interleaved runs/version, 256 requests/run, 128 input and 128 output tokens, concurrency 32) was intentionally neutral on the GPU-bound model: median output throughput 952.90 -> 953.55 tok/s (+0.07%, within noise).

## AI assistance disclosure

AI assistance was used to  implement the change, and prepare tests and benchmarks. The human submitter reviewed every changed line and the reported results and can explain and defend the change end-to-end.

### CODE REVIEW

File: vllm/v1/worker/gpu/sample/sampler.py [hunk 1]
```diff
@@ -229,22 +246,7 @@ def apply_sampling_params(
...
-        states = self.sampling_states
-        temperatures = states.temperature.np[idx_mapping_np]
-        if np.any((temperatures != 0.0) & (temperatures != 1.0)):
-            return True
-        if np.any(states.min_p.np[idx_mapping_np] != 0.0):
-            return True
-        if np.any(states.top_k.np[idx_mapping_np] != states.vocab_size):
-            return True
-        return bool(np.any(states.top_p.np[idx_mapping_np] != 1.0))
+        return bool(np.any(self._needs_logits_processing[idx_mapping_np]))
```
Comment (njhill) [inline vllm/v1/worker/gpu/sample/sampler.py]: Maybe remove this method now, can just check inline above
```python
        if not np.any(self.needs_logits_processing[idx_mapping_np]):
```

File: vllm/v1/worker/gpu/sample/sampler.py [hunk 2]
```diff
@@ -69,6 +70,22 @@ def add_request(
         self.logprob_token_ids_state.add_request(req_idx, sampling_params)
         self.thinking_budget_state.add_request(req_idx, sampling_params)
 
+        states = self.sampling_states
+        temperature = states.temperature.np[req_idx]
+        self._needs_logits_processing[req_idx] = (
```
Comment (njhill) [inline vllm/v1/worker/gpu/sample/sampler.py]: nit - let's avoid the underscore for consistency with other fields here
```suggestion
        self.needs_logits_processing[req_idx] = (
```

**Review by njhill — COMMENTED** [review]
Thanks @positive666 just a couple of small requests

File: vllm/v1/worker/gpu/sample/sampler.py (hunk 1 above)
Comment (positive666) [reply vllm/v1/worker/gpu/sample/sampler.py]: Done in 16713c99. I removed the single-use helper and inlined the active-request cache check in `apply_sampling_params()`.

File: vllm/v1/worker/gpu/sample/sampler.py (hunk 2 above)
Comment (positive666) [reply vllm/v1/worker/gpu/sample/sampler.py]: Done, renamed it for consistency. Thanks!

**Comment by positive666** [comment]
@njhill Addressed both comments and reran the focused GPU tests — 26 passed. Thanks!

**Review by njhill — APPROVED** [review]
Thanks @positive666
